File: frontend/src/services/borrowRequestService.js
import axiosInstance from '../lib/axios';

/**
 * @typedef {import('../types/borrowRequest').BorrowRequest} BorrowRequest
 * @typedef {import('../types/borrowRequest').CreateBorrowRequestDTO} CreateBorrowRequestDTO
 * @typedef {import('../types/borrowRequest').RequestStatistics} RequestStatistics
 * @typedef {import('../types/borrowRequest').RequestStatus} RequestStatus
 */

/**
 * Borrow Request service for handling the complete borrow workflow
 */
const borrowRequestService = {
  /**
   * Create a new borrow request
   * @param {number} itemId - Item ID to borrow
   * @param {CreateBorrowRequestDTO} data - Request data
   * @returns {Promise<{success: boolean, data?: BorrowRequest, message?: string, error?: string, status?: number, validationErrors?: Object}>} Response with created request data
   */
  async createRequest(itemId, data) {
    try {
      const response = await axiosInstance.post(`/requests?itemId=${itemId}`, data);
      return {
        success: true,
        data: response.data,
        message: 'Borrow request created successfully',
      };
    } catch (error) {
      return {
        success: false,
        error: error.response?.data?.message || 'Failed to create borrow request. Please try again.',
        status: error.response?.status,
        validationErrors: error.response?.data?.errors,
      };
    }
  },

  /**
   * Get all requests sent by the current user (borrower view)
   * @param {RequestStatus} [status] - Optional status filter
   * @returns {Promise<{success: boolean, data?: BorrowRequest[], error?: string, status?: number}>} Response with list of sent requests
   */
  async getSentRequests(status = null) {
    try {
      const params = status ? { status } : {};
      const response = await axiosInstance.get('/requests/sent', { params });
      return {
        success: true,
        data: response.data,
      };
    } catch (error) {
      return {
        success: false,
        error: error.response?.data?.message || 'Failed to fetch sent requests.',
        status: error.response?.status,
      };
    }
  },

  /**
   * Get all requests received by the current user (lender view)
   * @param {RequestStatus} [status] - Optional status filter
   * @returns {Promise<{success: boolean, data?: BorrowRequest[], error?: string, status?: number}>} Response with list of received requests
   */
  async getReceivedRequests(status = null) {
    try {
      const params = status ? { status } : {};
      const response = await axiosInstance.get('/requests/received', { params });
      return {
        success: true,
        data: response.data,
      };
    } catch (error) {
      return {
        success: false,
        error: error.response?.data?.message || 'Failed to fetch received requests.',
        status: error.response?.status,
      };
    }
  },

  /**
   * Get request by ID
   * @param {number} requestId - Request ID
   * @returns {Promise<{success: boolean, data?: BorrowRequest, error?: string, status?: number}>} Response with request details
   */
  async getRequestById(requestId) {
    try {
      const response = await axiosInstance.get(`/requests/${requestId}`);
      return {
        success: true,
        data: response.data,
      };
    } catch (error) {
      return {
        success: false,
        error: error.response?.data?.message || 'Failed to fetch request details.',
        status: error.response?.status,
      };
    }
  },

  /**
   * Approve a borrow request (lender only)
   * @param {number} requestId - Request ID
   * @param {string} [responseMessage] - Optional response message
   * @returns {Promise<{success: boolean, data?: BorrowRequest, message?: string, error?: string, status?: number}>} Response with updated request data
   */
  async approveRequest(requestId, responseMessage = null) {
    try {
      const data = responseMessage ? { responseMessage } : {};
      const response = await axiosInstance.post(`/requests/${requestId}/approve`, data);
      return {
        success: true,
        data: response.data,
        message: 'Request approved successfully',
      };
    } catch (error) {
      return {
        success: false,
        error: error.response?.data?.message || 'Failed to approve request.',
        status: error.response?.status,
      };
    }
  },

  /**
   * Reject a borrow request (lender only)
   * @param {number} requestId - Request ID
   * @param {string} [responseMessage] - Optional rejection reason
   * @returns {Promise<{success: boolean, data?: BorrowRequest, message?: string, error?: string, status?: number}>} Response with updated request data
   */
  async rejectRequest(requestId, responseMessage = null) {
    try {
      const data = responseMessage ? { responseMessage } : {};
      const response = await axiosInstance.post(`/requests/${requestId}/reject`, data);
      return {
        success: true,
        data: response.data,
        message: 'Request rejected',
      };
    } catch (error) {
      return {
        success: false,
        error: error.response?.data?.message || 'Failed to reject request.',
        status: error.response?.status,
      };
    }
  },

  /**
   * Mark item as returned (lender only)
   * @param {number} requestId - Request ID
   * @returns {Promise<{success: boolean, data?: BorrowRequest, message?: string, error?: string, status?: number}>} Response with updated request data
   */
  async markAsReturned(requestId) {
    try {
      const response = await axiosInstance.post(`/requests/${requestId}/return`);
      return {
        success: true,
        data: response.data,
        message: 'Item marked as returned',
      };
    } catch (error) {
      return {
        success: false,
        error: error.response?.data?.message || 'Failed to mark item as returned.',
        status: error.response?.status,
      };
    }
  },

  /**
   * Confirm return and complete transaction (borrower only)
   * @param {number} requestId - Request ID
   * @returns {Promise<{success: boolean, data?: BorrowRequest, message?: string, error?: string, status?: number}>} Response with updated request data
   */
  async confirmReturn(requestId) {
    try {
      const response = await axiosInstance.post(`/requests/${requestId}/confirm`);
      return {
        success: true,
        data: response.data,
        message: 'Return confirmed successfully',
      };
    } catch (error) {
      return {
        success: false,
        error: error.response?.data?.message || 'Failed to confirm return.',
        status: error.response?.status,
      };
    }
  },

  /**
   * Cancel a pending request (borrower only)
   * @param {number} requestId - Request ID
   * @returns {Promise<{success: boolean, message?: string, error?: string, status?: number}>} Response with success status
   */
  async cancelRequest(requestId) {
    try {
      await axiosInstance.delete(`/requests/${requestId}`);
      return {
        success: true,
        message: 'Request canceled successfully',
      };
    } catch (error) {
      return {
        success: false,
        error: error.response?.data?.message || 'Failed to cancel request.',
        status: error.response?.status,
      };
    }
  },

  /**
   * Get request statistics for the current user
   * @returns {Promise<{success: boolean, data?: RequestStatistics, error?: string, status?: number}>} Response with statistics data
   */
  async getStatistics() {
    try {
      const response = await axiosInstance.get('/requests/statistics');
      return {
        success: true,
        data: response.data,
      };
    } catch (error) {
      return {
        success: false,
        error: error.response?.data?.message || 'Failed to fetch statistics.',
        status: error.response?.status,
      };
    }
  },
};

export default borrowRequestService;

File: backend/src/main/java/com/rentkar/service/BorrowRequestService.java
package com.rentkar.service;

import com.rentkar.dto.BorrowRequestDTO;
import com.rentkar.dto.CreateBorrowRequestDTO;
import com.rentkar.model.BorrowRequest;
import com.rentkar.model.RequestStatus;
import com.rentkar.model.User;

import java.time.LocalDate;
import java.util.List;

public interface BorrowRequestService {
    
    /**
     * Create a new borrow request
     * @param itemId The ID of the item to borrow
     * @param dto The request data
     * @param borrower The user creating the request
     * @return The created borrow request DTO
     */
    BorrowRequestDTO createRequest(Long itemId, CreateBorrowRequestDTO dto, User borrower);
    
    /**
     * Get all requests sent by a user (borrower view)
     * @param borrower The user who sent the requests
     * @param status Optional status filter (null for all)
     * @return List of borrow requests
     */
    List<BorrowRequest> getSentRequests(User borrower, RequestStatus status);
    
    /**
     * Get all requests received by a user (lender view)
     * @param lender The user who owns the items
     * @param status Optional status filter (null for all)
     * @return List of borrow requests
     */
    List<BorrowRequest> getReceivedRequests(User lender, RequestStatus status);
    
    /**
     * Get request by ID with authorization check
     * @param id The request ID
     * @param user The user requesting access
     * @return The borrow request
     * @throws RuntimeException if request not found or user not authorized
     */
    BorrowRequest getRequestById(Long id, User user);
    
    /**
     * Approve a borrow request (lender only)
     * @param id The request ID
     * @param responseMessage Optional response message from lender
     * @param lender The user approving the request
     * @return The approved borrow request
     * @throws RuntimeException if request not found, not pending, or user not authorized
     */
    BorrowRequest approveRequest(Long id, String responseMessage, User lender);
    
    /**
     * Reject a borrow request (lender only)
     * @param id The request ID
     * @param responseMessage Optional reason message from lender
     * @param lender The user rejecting the request
     * @return The rejected borrow request
     * @throws RuntimeException if request not found, not pending, or user not authorized
     */
    BorrowRequest rejectRequest(Long id, String responseMessage, User lender);
    
    /**
     * Mark item as returned (lender only)
     * @param id The request ID
     * @param lender The user marking the item as returned
     * @return The updated borrow request
     * @throws RuntimeException if request not found, not approved, or user not authorized
     */
    BorrowRequest markAsReturned(Long id, User lender);
    
    /**
     * Confirm return and complete transaction (borrower only)
     * @param id The request ID
     * @param borrower The user confirming the return
     * @return The completed borrow request
     * @throws RuntimeException if request not found, not returned, or user not authorized
     */
    BorrowRequest confirmReturn(Long id, User borrower);
    
    /**
     * Cancel a pending borrow request (borrower only)
     * @param id The request ID
     * @param borrower The user canceling the request
     * @throws RuntimeException if request not found, not pending, or user not authorized
     */
    void cancelRequest(Long id, User borrower);
    
    /**
     * Get request statistics for a user
     * @param user The user to get statistics for
     * @return Request statistics including counts by status for sent and received requests
     */
    com.rentkar.dto.RequestStatistics getStatistics(User user);
}
